LANGUAGE.md

```diff
@@ -0,0 +1,114 @@
+# Language Reference Manual
+
+This file is not a tutorial, but describes the language with a high-level grammar.
+It also describes how names and scopes are handled, as well as sort checking.
+
+## Grammar
+
+A program is a single module. (In the future, multiple modules will be supported.)
+
+    program ::= module
+
+A module consists of a sequence of declarations, each of which either (1)
+adds to the signature, (2) defines a macro, or (3) states a theorem.
+
+    module ::= declaration*
+    declaration ::= signature_declaration | definition | theorem_statement
+
+### Signature declarations
+
+A signature is a list of (global, uninterpreted) sorts and functions.
+
+A sort is declared by giving its name. (Unlike some other logical systems,
+all sorts are first order (cannot take other sorts as arguments).)
+
+    signature_declaration ::= sort_declaration | function_declaration
+    sort_declaration ::= "sort" identifier
+
+A function is declared by giving its mutability, its name, the sorts of its
+arguments, and its return sort.
+
+    function_declaration ::= mutability identifier function_arguments ":" sort
+    mutability ::= "mutable" | "immutable"
+    function_arguments ::= "(" zero_or_more_separated(sort, ",")  ")"
+
+A sort is either `bool` or the name of an uninterpreted sort.
+
+    sort ::= "bool" | identifier
+
+We sometimes refer to functions that return `bool` as "relations", and functions
+that take zero arguments as "constants".
+
+An uninterpreted sort is not allowed to be named "`bool`".
+
+### Definitions
+
+A definition defines a macro by giving its name, the names and sorts of its
+arguments, its return sort, and its body.
+
+    definition ::= "def" ident definition_arguments "->" sort "{" term "}"
+    definition_arguments ::= "(" zero_or_more_separated(definition_argument, ",") ")"
+    definition_argument ::= ident ":" sort
+
+Terms are described below.
+
+### Theorem statements
+
+A theorem statement is either an `assume` or an `assert`, each of which takes a
+term. An `assert` can optionally provide a `proof`, which is a sequence of
+`invariant` terms.
+
+    theorem_statement ::= assume_statement | assert_statement
+    assume_statement ::= "assume" term
+    assert_statement ::= "assert" term proof?
+    proof ::= "proof" "{" invariant* "}"
+    invariant ::= "invariant" term
+
+### Terms
+
+TODO
+
+## Scope and sort checking
+
+### Scopes
+
+- There are two global scopes: the sort scope and the function-and-definition
+  scope.
```

Comment:
Here as well let's just describe the current implementation.

Reply:
This already describes the current implementation, but that's because one of the reasons I started writing this file was to design the implementation with Alex :)

I don't feel two strongly about whether we should move to one scope in the future. I would be fine either way. One argument for having two scopes is that it lets you name a variable the same name as its type, as in

```
sort node
mutable r(node): bool
assert forall node. r(node)
```

I think that's kind of a convenient idiom, but it's not a huge deal to disallow it.